### PR TITLE
Preserve indent when commenting

### LINKED ISSUE #33069: Commenting in code editor to preserve indenting
**Describe the bug**
As a user I want to be able to comment out code in a loop or branch using the keyboard shortcut `ctrl+/`, however, currently this does not preserve the indenting (as per PyCharm)

**To Reproduce**
Trying to comment out the second line in this code produces
```
for ii in range(2):
#     print(ii)
    print(2*ii)
```
**Expected behavior**
The above still works but would be clearer if it was like

```
for ii in range(2):
    # print(ii)
    print(2*ii)
```

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Preserve indent when commenting with `ctrl-/` to give mantid same behaviour as PyCharm.

**To test:**
(1) Copy this into the script editor
```
for ii in range(2):
    print(ii)
print(10*ii)
```
(2) Comment lines with `ctrl-/` - it should produce this
```
# for ii in range(2):
    # print(ii)
# print(10*ii)
```
(3) Uncomment the lines and check the formatting is correct
(4) Play around with adding trailing whitespace etc.

Fixes #33069 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by thomashampson — CHANGES_REQUESTED** [review]

File: qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_codecommenter.py [hunk 1]
```diff
@@ -75,19 +77,28 @@ def test_multiline_uncomment(self):
...
+            '# ' + line for line in expected_lines[start_line:end_line + 1]]
         self.assertEqual(self.editor.text(), '\n'.join(expected_lines))
 
+    def test_multiline_comment_preserves_indenting(self):
+        start_line, end_line = 10, 13
+        self.editor.setSelection(start_line, 5, end_line, 5)
+        self.commenter.toggle_comment()
+        expected_lines = ["# for ii in range(2):",
+                          "   # do_something()",
+                          "# do_something_else()"]
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_codecommenter.py:99]: Thanks, actually I'd not realised PyCharm behaved like that - but I will aim for that now!

File: docs/source/release/v6.3.0/mantidworkbench.rst [hunk 2]
```diff
@@ -42,5 +42,6 @@ Bugfixes
 - Workbench will no longer hang if an algorithm was running when workbench was closed.
 - Stopped workbench from ignoring GUIs that want to cancel closing
 - Fixed a bug in the editor where uncommenting using 'ctrl+/' wasn't working correctly for lines of the form '<optional whitespace>#code_here # inline comment'.
+- Commenting code int he editor using 'ctrl+/' will preserve indenting (i.e. `# ` will be inserted at the position of the first non-whitespace character in the line).
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply docs/source/release/v6.3.0/mantidworkbench.rst]: Thanks, fixed this!

File: qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_codecommenter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply qt/python/mantidqt/mantidqt/widgets/codeeditor/test/test_codecommenter.py]: I think actually (I didn't realise this) but PyCharm uses the least indented line to determine the indentation of the comment - e.g. commenting this
```
    lines[i] = lines[i][:istart] + '# ' + lines[i][istart:]
return lines
```
gives this
```
#     lines[i] = lines[i][:istart] + '# ' + lines[i][istart:]
# return lines
```
So I now loop over all lines to determine the position of the first non-whitespace char with the minimum indent in the selection

**Review by thomashampson — APPROVED** [review]
This is a nice improvement to the way code is commented out when using the shortcut key, and works well for all the things I've tested (seems to be consistent with PyCharm's behaviour).

**Pull request merged**